src/coreComponents/mesh/utilities/MeshMapUtilities.hpp

```diff
@@ -126,103 +131,167 @@ inline localIndex size1( ArrayOfSetsView< T const > const & map, localIndex cons
+ * @param comp comparison function, must be copyable and compatible with @p POLICY
+ * @return the resulting ArrayOfSets
  */
-template< typename T >
-GEOSX_HOST_DEVICE
-inline localIndex numElements( arrayView1d< arrayView1d< T const > const > const & map )
+template< typename POLICY, typename T, typename COMP = LvArray::sortedArrayManipulation::less< T > >
+ArrayOfSets< T >
+makeArrayOfSets( ArrayOfArrays< T > && input, bool valuesAreUnique, COMP comp = {} )
 {
-  return map.size();
-}
+  // Pre-sort each sub-array in parallel (to avoid currently serial sorting in assimilate() call)
+  if( valuesAreUnique )
+  {
+    forAll< POLICY >( input.size(), [comp, view = input.toView()] ( localIndex const i )
+    {
+      arraySlice1d< T > const row = view[i];
+      LvArray::sortedArrayManipulation::makeSorted( row.begin(), row.end(), comp );
+    } );
+  }
+  else
+  {
+    forAll< POLICY >( input.size(), [comp, view = input.toView()] ( localIndex const i )
+    {
+      arraySlice1d< T > const row = view[i];
+      LvArray::sortedArrayManipulation::makeSortedUnique( row.begin(), row.end(), comp );
+    } );
+  }
 
-/**
- * @copydoc numElements(arrayView1d< arrayView1d< T const > const > const &)
- * @tparam USD unit-stride dimension of the map
- */
-template< typename T, int USD >
-GEOSX_HOST_DEVICE
-inline localIndex numElements( arrayView2d< T, USD > const & map )
-{
-  return map.size();
+  // Move the temporary storage into the array of sets
+  ArrayOfSets< T > result;
+  result.assimilate( std::move( input ), LvArray::sortedArrayManipulation::SORTED_UNIQUE );
+  return result;
 }
 
 /**
- * @copydoc numElements(arrayView1d< arrayView1d< T const > const > const &)
+ * @brief Build to-element map by inverting existing maps in element subregions.
+ * @tparam FUNC type of @p cellToObjectGetter
+ * @param elementRegionManager element region manager
+ * @param numObjects number of objects (nodes, faces, etc.) for which maps are built
+ * @param toElements container for to-element (region, subregion, index) maps
+ * @param cellToObjectGetter function used to extract maps from subregions
+ * @param overAlloc overallocation for the maps (extra capacity per row)
  */
-template< typename T >
-GEOSX_HOST_DEVICE
-inline localIndex numElements( ArrayOfArraysView< T const > const & map )
+template< typename FUNC >
+void buildToElementMaps( ElementRegionManager const & elementRegionManager,
+                         localIndex const numObjects,
+                         OrderedVariableToManyElementRelation & toElements,
+                         FUNC cellToObjectGetter,
+                         localIndex const overAlloc = 0 )
 {
-  localIndex size = 0;
-  for( localIndex i = 0; i < map.size(); ++i )
+  ArrayOfArrays< localIndex > & toElementRegionList = toElements.m_toElementRegion;
+  ArrayOfArrays< localIndex > & toElementSubRegionList = toElements.m_toElementSubRegion;
+  ArrayOfArrays< localIndex > & toElementList = toElements.m_toElementIndex;
```

Comment:
Sure, it's up for debate, there's more than one way to approach building these maps.

My take is, `m_toElementIndex` was a questionable input. First of all, it's redundant w.r.t. element-to-node maps that already exist in subregions. Second, it contained *some* element indices which were neither global nor unique - they were element indices within *some* cell block, without specifying which one: for some node, you could have an element list like `[5,5,5,5]`, which is just weird? Indeed, we later establish which cell block it was by traversing element-to-node maps, but if we're required to traverse them anyway, why not just build the whole `toElement` connectivity from it?

In other words, I think the previous version was doing more work than it needed to:
  * first, traverse element-to-node maps of each `CellBlock` in order to build `toElementList` but lose block indices
  * then, traverse element-to-node maps again (now of each `SubRegion`), but while visiting each node, also search through its `toElementList` to try and find current element in it and thus recover its (sub)region index. This is an extra inner loop, which doesn't change asymptotic complexity (it's bounded by the secondary size of node-to-element adjacency, and I admit constant factors in O-notation are hard to reason about without direct measurement), but still it's extra work that could be avoided.

As a second concern, I couldn't see a straightforward way to parallelize the existing algorithm. The inner loop and the tricks in it (like inserting a bunch of `-1`s) made it difficult to use atomics which are required since the outer parallel loop is over elements. While just function was **not** a bottleneck of map generation, I prefer not to leave any serial code when dealing with mesh things that may scale. The original code by @corbett5 was also not parallelized, but it contained a comment about doing it via tuples, which is what I implemented. It is now slower in serial because of the extra memory bandwidth involved (building a temporary structure with tuples), but I'm almost certain there is a (realistic) mesh size and number of threads at which it is faster.

Reply:
> First of all, it's redundant w.r.t. element-to-node maps that already exist in subregions.

What do you mean by redundant? `CellElementSubRegion` provide their own `index` plus some `element -> node` mappings and this is the inverted relationship? Wouldn't this argument apply to all mappings that we invert?

What I'm OK with is to unite `node` to `elem` mapping and `CellElementSubRegion` indices because these can be considered as _raw meshing concepts_. What I'm far less comfortable with is doing the `ElementRegion` in the same time, because it's not geometrical.

Another issue is that your PR is changing the contract of the `CellBlockManagerABC` without doing it.
I'm OK to make changes but those changes need to be consistent.

Right now, we still have `CellBlockManagerABC::getNodeToElements()` which is used in `NodeManager::setGeometricalRelations( const CellBlockManagerABC & )` to define `m_toElements.m_toElementIndex = cellBlockManager.getNodeToElements()`.
But this action is now useless since `m_toElements.m_toElementIndex` is overwritten in `NodeManager::buildRegionMaps( const ElementRegionManager & )`.

So who's responsible for computing `m_toElements.m_toElementIndex`, `m_toElements.m_toElementSubRegion` and `m_toElements.m_toElementRegion`? The situation of the current PR does not settle things for sure and that makes me uncomfortable. I would advocate for 
- Pure geometry goes to `CellBlockManangerABC` side.
- Simulation information is done in another place.

Reply:
> What do you mean by redundant? `CellElementSubRegion` provide their own `index` plus some `element -> node` mappings and this is the inverted relationship? Wouldn't this argument apply to all mappings that we invert?

Redundant as in, not containing any information that is not available from another data structure. In case of nodes, each `CellElementSubRegions`'s elem-to-node map (which is built by cell block and copied over) acts as the single source of truth about geometrical relationship, and node-to-element map is built just as an alternative representation of the same data used for convenience and speed of access.

And indeed, the same applies to e.g. face-to-elem map, but because it's stored differently (array2d), I haven't yet found a satisfactory way to invert elem-to-face in parallel (perhaps building a temporary `ArrayOfArrays` first, using the same function `buildElementMaps()`, and then copying into `array2d` is the way to go).

> Another issue is that your PR is changing the contract of the CellBlockManagerABC without doing it.
I'm OK to make changes but those changes need to be consistent.
> So who's responsible for computing m_toElements.m_toElementIndex, m_toElements.m_toElementSubRegion and m_toElements.m_toElementRegion?

Ideally, I would have `CellBlockManager` classes build only downward-looking maps (`cell-to-[face/edge/node]`, `face-to-[edge/node]`, `edge-to-node`), and recover the up-looking maps by inverting them on the `MeshLevel` side of the interface, as needed. It's a pretty clean contract, in my opinion.

The downside of this approach is that we lose the ability to specialize the algorithms for the up-map construction, but if the generic algorithm is fast enough, maybe that's not too big of a deal.

> Right now, we still have CellBlockManagerABC::getNodeToElements() which is used in NodeManager::setGeometricalRelations( const CellBlockManagerABC & ) to define m_toElements.m_toElementIndex = cellBlockManager.getNodeToElements().
But this action is now useless since m_toElements.m_toElementIndex is overwritten in NodeManager::buildRegionMaps( const ElementRegionManager & ).

It's an oversight for sure, I meant to delete the line 
```c++
m_toElements.m_toElementIndex = cellBlockManager.getNodeToElements();
```
(which leaves `getNodeToElements()` unused except for a single unit-test assertion, so I would propose to remove it)

Reply:
> Ideally, I would have `CellBlockManager` classes build only downward-looking maps (`cell-to-[face/edge/node]`, `face-to-[edge/node]`, `edge-to-node`), and recover the up-looking maps by inverting them on the `MeshLevel` side of the interface, as needed. It's a pretty clean contract, in my opinion.
> 
> The downside of this approach is that we lose the ability to specialize the algorithms for the up-map construction, but if the generic algorithm is fast enough, maybe that's not too big of a deal.

Hi @jeannepellerin do you have any opinion about that?
Could this approach (in a nutshell, inverting the geometrical maps outside of the `CellBlockManager` zone) be compatible with your developments?
It could be fruitful to get from your expertise before we modify this architecture component.

@klevzoff Does it make sense to compute the `m_toElements.m_toElementIndex`, `m_toElements.m_toElementSubRegion` upstream while computing `m_toElements.m_toElementRegion` later on? I would expect this to be less a problem but I'm not sure?

Reply:
Well, I don't insist we necessarily do this, and this PR does not touch the inversion any maps other than `toElement` ones - in fact, the way some of these upward maps are being constructed currently is more efficient than an explicit inversion. It was just my proposal for a possible contract for `CellBlockManagerABC`, which allows it to do most of the heavy lifting - such as identification of unique faces and edges - and return the processed connectivity information in non-redundant form.

> The question is do we want the mappings to be independant from one another?

No, in my opinion. E.g. `elementToNode` and `nodeToElement` are not independent, they contain the exact same information (in different forms) and must be consistent with each other 100% of the time. Having a unique point of construction of `nodeToElement` (rather than delegating it to multiple `CellBlockManagerABC` implementations) aids in that goal. If there is an efficiency argument for the construction of `nodeToElement` to be handled by `CellBlockManager` rather than `NodeManager`, that's a different story. With the current `develop` version that is not the case, hence my changes.

> @klevzoff Does it make sense to compute the `m_toElements.m_toElementIndex`, `m_toElements.m_toElementSubRegion` upstream while computing `m_toElements.m_toElementRegion` later on? I would expect this to be less a problem but I'm not sure?

`CellBlockManager` cannot construct `m_toElementSubRegion`, because it has no idea about subregion indices, only block indices (which are different!). It could potentially provide a
```c++
virtual std::pair< ArrayOfArrays< localIndex >, ArrayOfArrays< localIndex > >
getNodeToCellAndBlock();
```
kind of API, but in order to efficiently turn it into the final form, we also need `blockIndex -> (regionIndex, subRegionIndex)` mapping that is currently missing (but could be constructed).

Reply:
> ... elementToNode and nodeToElement are not independent, they contain the exact same information (in different forms) and must be consistent with each other 100% of the time. Having a unique point of construction of nodeToElement (rather than delegating it to multiple CellBlockManagerABC implementations) aids in that goal.

I do agree with you that consistency is required.
_One way_ to get this is to impose a map transposition for everyone.
But
- The less you see the better it is, so we could transpose all the mappings far away from the managers (which is basically what we are doing right now for most of the mappings, I presume)
- It may not be the _only way_. For seismic propagation where there can be very large cartesian grids, a mapping and its inverse can surely be computed "_analytically_" (assuming that we need the two mappings, which I do not know about).

> CellBlockManager cannot construct m_toElementSubRegion, because it has no idea about subregion indices, only block indices (which are different!).

When I did my #1418 refactoring (predestined id...) I think that I did _at least_ the two following questionable choices:
- The dynamic arrays in the computation to prepare _Corner Point Grids_. It was not the best idea and I'm glad we move backwards on this. There will be long since we support _CPG_ so no need to burden the performances of everyone. I already told about it and it's not the core subject here.
- The computation of the `m_toElements` triple mappings: `nodes`to `element index`, `sub-region index`, `region index`. I wanted to exclude the computation of the `region index` because I considered that regions are not a computational geometry concept but zones to define where to apply some physics. I decided not to compute the `sub-region` mappings behind the `CellBlockManagerABC` curtain, also because we do not really know about these `sub-regions` at that point. I've let this completely outside of the `CellBlockManager` zone and I did perform _no intermediate computation regarding these mappings_. But like for the first point, the current discussion makes me think that this cannot be a final answer (_vide infra_).

> It could potentially provide a
> ```c++
> virtual std::pair< ArrayOfArrays< localIndex >, ArrayOfArrays< localIndex > > getNodeToCellAndBlock();
> ```
> kind of API

`CellBlockMangerABC` does know about `SubRegions` and this is fine.

It can be enriched and provide a `getNodeToBlock()` alongside the `getNodesToElements()` (probably what you mean with the `std::pair`).

I'd expect that those two mappings could be computed in the same efficient way as you currently do (except that we would not compute the `node to region` mapping).
Then it would be the duty of the `ElementRegionManger` which manipulates the `SubRegions` and the `CellBlocks` to transform the mappings consistently (maybe this can be done in-place?). If `ElementRegionManager` wants to keep the same indexing, great.
And I still don't know about the `node to region` mapping, I naively hope it would not be very complicated to build apart but I'n not sure.

Reply:
@jeannepellerin Upon further discussion with @TotoGaz, we are leaving all maps and responsibilities where they are in `develop` (and your PR). The only change I will implement is this: in addition to `noteToCell` map, `CellBlockManager` will return `nodeToBlock` map to provide the complete information about this connectivity (same goes for faces and possibly edges). This map in each entry just contains the block index corresponding to the cell index in `nodeToCell`, but otherwise has the same shape and can be built simultaneously.

Reply:
OK. Yes, this mapping to Blocks is needed. I do not know how the clients of the class were supposed to guess the CellBlock.
These mapping are all strongly interdependant, but we can make their computations independant from one another, not all of them for sure. And a very fast map inversion is definitely the best way to do some of them.

Reply:
> ... elementToNode and nodeToElement are not independent, they contain the exact same information (in different forms) and must be consistent with each other 100% of the time. Having a unique point of construction of nodeToElement (rather than delegating it to multiple CellBlockManagerABC implementations) aids in that goal.

Oh, and I forgot to mention the most obvious use case 😵‍💫 (listed in https://github.com/GEOSX/GEOSX/issues/1762): "_Implement a PrecomputedCellBlockManager that would read data from disk and could prevent from recomputing everything every time._". If everything is precomputed, and read from disk, no need to recompute the mappings! This could be a great feature for a lot of users or researchers: when you want to test numerical strategies, you do not want to recompute the meshing part too many times.

> OK. Yes, this mapping to Blocks is needed. I do not know how the clients of the class were supposed to guess the CellBlock.

Funambulist algorithm there 🙄 🤣